backend/src/youtube_handler.py

```diff
@@ -0,0 +1,136 @@
+from  googleapiclient import discovery
+from secrets_handler import SecretsHandler
+import queue
+import time
+import threading
+from googleapiclient.errors import HttpError
+
+
+class YoutubeHandler:
+    """
+    A handler for interacting with the YouTube API.
+
+    Contains a definition for a static _client_instance. This should only be
+    accessed via the `youtube` field, which will automatically fill this field
+    if it does not yet exist. Any other accesses to this instance are unsafe
+    and should not be used.
+
+
+    """
+    api_service_name = "youtube"
+    api_version = "v3"
+    MAX_THREAD_COUNT: int = 5
+    _youtube_instance: discovery.Resource | None = None
+    request_queue: queue.Queue = queue.Queue()
+    thread_pool: list[threading.Thread] = []
+
+
+    def __init__():
+        """
+        Raises a TypeError.
+        YoutubeHandler instances must not be created as this is a singleton
+        """
+        raise TypeError(
+            "YoutubeHandler instances should not be created. " +
+            "Consider using the `get_client()` function."
+        )
+    @staticmethod
+    def get_client():
+        """
+        Retrieves or creates the YouTube client instance
+
+        Returns
+        -------
+        googleapiclient.discovery.Resource
+            The static YouTube client
+        """
+        if YoutubeHandler._youtube_instance:
+            return YoutubeHandler._youtube_instance
+
+        return YoutubeHandler._initialize_client()
+    @staticmethod
+    def _initialize_client():
+        """
+        Initializes the YouTube client.
+        """
+        _youtube_instance = discovery.build(
+            YoutubeHandler.api_service_name,
+            YoutubeHandler.api_version,
+            developerKey=SecretsHandler.get_youtube_key()
+        )
+        return _youtube_instance
+    
+    @staticmethod
+    def search_for_match(song: dict):
```

Comment:
This'll need a typehint for what type of dict is expected for `song` (dict[str, str] for example), but I do realize that the song dictionary is huge and has tons of types in it. I think we define a `songs_t` type in SpotifyHandler or in a types.py file maybe

Reply:
actually a great idea, I was getting the feeling this dict needed more specific typing, but it didn't even come to mind to just make a new type like that, will be doing